source/lubeck.d

```diff
@@ -1025,3 +1050,1019 @@ unittest
+              .sliced(4, 4)
+              .as!double.slice;
+    auto QR = A.qr;
+    auto res = mtimes(QR.qt.transposed, QR.rt.transposed).universal;
+    res = res.transposed;
+
+    import mir.ndslice.algorithm: equal;
+    assert(equal!((a, b) => fabs(a - b) < 1e-12)(res, A));
+}
+
+unittest
+{
+    auto A =
+            [ 1,  1,
+              1,  0 ]
+              .sliced(2, 2)
+              .as!double.slice;
+    auto QR = A.qr;
+    auto res = mtimes(QR.qt.transposed, QR.rt.transposed).universal;
+    res = res.transposed;
+
+    import mir.ndslice.algorithm: equal;
+    assert(equal!((a, b) => fabs(a - b) < 1e-12)(res, A));
+}
+
+struct choleskyResult(T)
+{
+    ///If uplo = 'L': lower triangle of 'matrix' is stored.
+    ///If uplo = 'U': upper triangle of 'matrix' is stored.
+    char uplo;
+    ///if uplo = Lower, the leading 'N x N' lower triangular part of A
+    ///contains the lower triangular part of the matrix A, and the
+    ///strictly upper triangular part if A is not referenced.
+    ///if uplo = Upper, the leading 'N x N' upper triangular part of A
+    ///contains the upper triangular part of the matrix A, and the
+    ///strictly lower triangular part if A is not referenced.
+    Slice!(Canonical, [2], T*) matrix;
+    ///Return solves a system of linear equations A * X = B, using Cholesky factorization.
+    auto solve(Flag!"allowDestroy" allowDestroy = No.allowDestroy,
+               SliceKind kind, size_t[] n, Iterator)
+              (Slice!(kind, n, Iterator) b)
+    {
+        return choleskySolve!(allowDestroy)(matrix, b, uplo);
+    }
+}
+/++
+Computs Cholesky decomposition of symmetric positive definite matrix 'A'.
+The factorization has the form:
+    A = U**T * U, if UPLO = Upper, or
+    A = L * L**T, if UPLO = Lower
+Where U is an upper triangular matrix and L is lower triangular.
+Params:
+    a = symmetric 'N x N' matrix.
+    uplo = if uplo is Upper, then upper triangle of A is stored, else
+    lower.
+Returns: $(LREF choleskyResult)
++/
+auto choleskyDecomp(Flag!"allowDestroy" allowDestroy = No.allowDestroy,
+              SliceKind kind, Iterator)
+             (Slice!(kind, [2], Iterator) a,
+              char uplo)
+in
+{
+    assert(a.length!0 == a.length!1, "matrix must be squared");
+}
+body
+{
+    import mir.ndslice.algorithm: eachUploPair;
+    alias T = BlasType!Iterator;
+    if(allowDestroy && a._stride!1)
+    {
+        auto m = a.assumeCanonical;
+        potrf!T(m, uplo);
+        if(uplo == 'L')
+            m.eachUploPair!"a = 0";
+        else
+            m.eachUploPair!"b = 0";
+        return choleskyResult!T(uplo, m);
+    }
+    else
+    {
+        auto m = a.as!T.slice.canonical;
+        potrf!T(m, uplo);
+        if(uplo == 'L')
+            m.eachUploPair!"a = 0";
+        else
+            m.eachUploPair!"b = 0";
+        return choleskyResult!T(uplo, m);
+    }
+}
+
+/++
+    Solves a system of linear equations A * X = B with a symmetric matrix A using the
```

Comment:
As a general comment on API design, I'm not sure why we have separate functions for several different types of solvers. Personally, I'm a fan of the approach taken in `mir.math.sum` that uses a `Summation` enum to switch between the different algorithms. The feature that I would want is solving a set of linear equations. Choosing how to do that is an implementation detail. I shouldn't need to do something like
```
auto C = A.choleskyDecomp('L');
auto X = C.solve(B);
```
I would prefer
```
auto C = A.solve!("cholesky");
```